pkg/metrics/process.go

```diff
@@ -0,0 +1,249 @@
+			Help: "Process per request duration in seconds.",
+		},
+		[]string{"hostname", "pid", "pid_name", "ppid", "trace_id"},
+	)
+)
+
+func InitProcessMetricsServer(listenAddress string) <-chan error {
+	if listenAddress == "" {
+		listenAddress = os.Getenv("PROCESS_METRICS_LISTEN_ADDRESS")
+		if listenAddress == "" {
+			listenAddress = ":9090"
+		}
+	}
+	customRegistry := prometheus.NewRegistry()
+	customRegistry.MustRegister(ProcessCPUPercent)
+	customRegistry.MustRegister(ProcessMemoryPercent)
+	customRegistry.MustRegister(ProcessMemoryRSS)
+	customRegistry.MustRegister(ProcessRequestDuration)
+	customRegistry.MustRegister(ProcessPerRequestDuration)
+
+	var errCH chan error
+	go func() {
+		http.Handle("/process/metrics", promhttp.HandlerFor(customRegistry, promhttp.HandlerOpts{}))
+		errCH <- http.ListenAndServe(listenAddress, nil)
+	}()
+	return errCH
+}
+
+func RecordProcessResourceUsageMetrics(pid int32) error {
+	pi, err := CollectProcessInfo(pid)
+	if err != nil {
+		return err
+	}
+	hostName, _ := os.Hostname()
+	labels := prometheus.Labels{
+		"hostname": hostName,
+		"pid":      fmt.Sprintf("%v", pi.Pid),
+		"pid_name": pi.ProcessName,
+		"ppid":     fmt.Sprintf("%v", pi.ParentPid),
+	}
+	ProcessCPUPercent.With(labels).Set(pi.CPUInfo.CPUPercent)
+	ProcessMemoryPercent.With(labels).Set(float64(pi.MemoryInfo.MemoryPercent))
+	ProcessMemoryRSS.With(labels).Set(float64(pi.MemoryInfo.RSS))
+	return nil
+}
+
+func RecordProcessRequestMetrics(pid int32, traceID string, duration float64) error {
+	pi, err := CollectProcessInfo(pid)
+	if err != nil {
+		return err
+	}
+	hostName, _ := os.Hostname()
+	labels := prometheus.Labels{
+		"hostname": hostName,
+		"pid":      fmt.Sprintf("%v", pi.Pid),
+		"pid_name": pi.ProcessName,
+		"ppid":     fmt.Sprintf("%v", pi.ParentPid),
+	}
+	ProcessRequestDuration.With(labels).Observe(duration)
+	labels["trace_id"] = traceID
+	ProcessPerRequestDuration.With(labels).Set(duration)
+	return nil
+}
+
+type ProcessInfo struct {
+	Pid         int32
+	ParentPid   int32
+	ProcessName string
+	CPUInfo     *CPUInfo
+	MemoryInfo  *MemoryInfo
+}
+
+type CPUInfo struct {
+	CPUPercent float64
+}
+
+type MemoryInfo struct {
+	MemoryPercent float32
+	*process.MemoryInfoStat
+}
+
+func CollectProcessInfo(pid int32) (*ProcessInfo, error) {
+	p, err := process.NewProcess(pid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create process[%v] instance: %v\n", pid, err)
+	}
+	name, err := p.Name()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get process[%v] name: %v", pid, err)
+	}
+	parentPid, err := p.Ppid()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get process[%v] parent process pid: %v", pid, err)
+	}
+	cpuPercent, err := p.CPUPercent()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get process[%v] cpu percent: %v", pid, err)
+	}
+	memoryPercent, err := p.MemoryPercent()
+	if err != nil {
+		fmt.Errorf("failed to get process[%v] memory percent: %v", pid, err)
```

Comment:
The result of `fmt.Errorf` seems not used, should it be `return nil, fmt.Errorf()`?

Reply:
fixed